Automattic/vip-go-mu-plugins
# Add/mime type caching

## Description
This PR adds the logic for caching the MIME types by relying in the `pre_get_available_post_mime_types()` filter that will be added in 6.4 (https://core.trac.wordpress.org/ticket/52759).

**The code here is not final and needs additional testing for adding, updating, and removing attachments. I'm submitting is as a draft PR to start the conversation.**

Notes:
- The code will only run if the WP version is `>=` than `MINIMUM_WORDPRESS_VERSION`, which is set to 6.4.
- The effects of the code (and the caching) can be disabled altogether by ~~adding a `vip_cache_mime_types` filter and returning `false`~~ removing the `vip_cache_mime_types` action from `init`.
- Since the MIME type query could actually timeout for sites with a huge number of assets, I've included the constant `MAX_POSTS_TO_QUERY_DEFAULT` that will be used to determine if the querying/caching should be enabled or if it should fall back to the default WordPress MIME types, bringing the dropdown content to parity with the grid view. This can also be filtered via `vip_max_posts_to_query_for_mime_type_caching`.

To discuss:
- ~~Another simplified approach that would reduce the complexity would be to simply delete the cache on adding, updating, and deleting attachments and let it rebuild by itself the next time.~~ Edit: The cache is now being deleted.
- ~~If updating the caches as the code is doing it right now, we could also avoid impacting the attachment update and delete times (since an additional query might need to happen) by refactoring it to run the cache updates on shutdown.~~ Edit: This no longer applies as the cache is now being deleted on delete/update.
- ~~We could also leverage the cache updates to a cron process so we remove the update logic from the web.~~ Edit: This could be revisited later if needed.

## Changelog Description
### Feature added: MIME types caching for the media library

Cache MIME types via the `pre_get_available_post_mime_types()` filter introduced in WordPress 6.4.

## Pre-review checklist

Please make sure the items below have been covered before requesting a review:


## Pre-deploy checklist


## Steps to Test

1. Check out PR.
1. Go to `wp-admin` > `Media Library`
1. Click on the list view.
1. Verify that the attachment filter dropdown populates correctly for a site with a number of attachments less or equal to the `MAX_POSTS_TO_QUERY_DEFAULT` constant.
1. If the number of attachments exceeds the constant or the filtered value from `vip_max_posts_to_query_for_mime_type_caching`, then validate that the default attachment types are loaded. Its contents should match the dropdown from the grid view.
1. Validate that the items available in the dropdown also update accordingly when adding, updating, or deleting attachments.
